src/main/java/com/epam/aidial/core/token/TokenStatsTracker.java

```diff
@@ -2,22 +2,134 @@
 
 import com.epam.aidial.core.ProxyContext;
 import com.epam.aidial.core.config.ApiKeyData;
+import com.epam.aidial.core.util.ProxyUtil;
 import com.google.common.annotations.VisibleForTesting;
 import io.vertx.core.Future;
 import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import lombok.extern.slf4j.Slf4j;
+import org.redisson.api.RTopic;
+import org.redisson.api.RedissonClient;
+import org.redisson.client.codec.StringCodec;
 
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.LongSupplier;
 
+@Slf4j
 public class TokenStatsTracker {
+
+
     private final Map<String, TraceContext> traceIdToContext = new ConcurrentHashMap<>();
 
-    public void startSpan(ProxyContext context) {
+    /**
+     * List of registered trackers in the cluster
+     */
+    private final Map<String, TrackerData> registeredTrackers = new ConcurrentHashMap<>();
+
+    private final RTopic topic;
+
+    private final Vertx vertx;
+
+    private final LongSupplier clock;
+
+    private final String trackerId;
+
+    public TokenStatsTracker(RedissonClient redis, Vertx vertx, LongSupplier clock) {
+        this.clock = clock;
+        this.vertx = vertx;
+        this.topic = redis.getTopic("deployment-cost-stats-event", new StringCodec());
+        topic.addListener(String.class, (channel, event) -> onEvent(event));
+        this.trackerId = UUID.randomUUID().toString();
+        vertx.setPeriodic(0, 60_000, event -> cleanupDeadTrackers());
+        vertx.setPeriodic(0, 10_000, event -> ping());
+    }
+
+    private void cleanupDeadTrackers() {
+        long currentTime = clock.getAsLong();
+        for (Map.Entry<String, TrackerData> entry : registeredTrackers.entrySet()) {
+            TrackerData trackerData = entry.getValue();
+            String trackerId = entry.getKey();
+            if (currentTime - trackerData.lastUpdateTs > 60_000) {
+                // complete all promises forcibly
+                trackerData.completePromises();
+                registeredTrackers.remove(trackerId);
```

Comment:
race? You can complete/clear all promises, then addSpan can add one more, only then you will remove and the promise will never end

Reply:
Every tracker has own unique ID if the tracker is dead the code completes all promises. A new instance with a **different** ID can add more span.

Reply:
Right. The block should be synchronized. I'll fix it.

Reply:
fixed